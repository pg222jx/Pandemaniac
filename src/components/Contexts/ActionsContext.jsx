import React,{ useState, createContext } from 'react';

export const ActionsContext = createContext()

export const ActionsProvider = props => {

    const [amount, setAmount] = useState(2)

    return (
        <ActionsContext.Provider value={[amount, setAmount]}>
            {props.children}
        </ActionsContext.Provider>
    )
}

