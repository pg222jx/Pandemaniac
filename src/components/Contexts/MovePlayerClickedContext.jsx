import React,{ useState, createContext } from 'react';

export const MovePlayerClickedContext = createContext()

export const MovePlayerClickedProvider = props => {

    const [clicked, setClicked] = useState(false)

    return (
        <MovePlayerClickedContext.Provider value={[clicked, setClicked]}>
            {props.children}
        </MovePlayerClickedContext.Provider>
    );
}

