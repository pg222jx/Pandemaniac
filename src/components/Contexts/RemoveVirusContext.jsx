import React,{ useState, createContext } from 'react';

export const RemoveVirusContext = createContext()

export const RemoveVirusProvider = props => {

    const [clicked, setClicked] = useState(false)

    return (
        <RemoveVirusContext.Provider value={[clicked, setClicked]}>
            {props.children}
        </RemoveVirusContext.Provider>
    );
}

