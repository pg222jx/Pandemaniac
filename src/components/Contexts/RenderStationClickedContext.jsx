import React,{ useState, createContext } from 'react';

export const RenderStationClickedContext = createContext()

export const RenderStationClickedProvider = props => {

    const [clicked, setClicked] = useState(false)

    return (
        <RenderStationClickedContext.Provider value={[clicked, setClicked]}>
            {props.children}
        </RenderStationClickedContext.Provider>
    );
}

