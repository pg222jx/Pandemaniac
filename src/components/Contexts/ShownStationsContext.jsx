import React, { useState, createContext } from 'react'

export const ShownStationsContext = createContext()

export const ShownStationsProvider = props => {

    const [shownStations, setShownStations] = useState([])

    return (
        <ShownStationsContext.Provider value={[shownStations, setShownStations]}>
            {props.children}
        </ShownStationsContext.Provider>
    )
}

