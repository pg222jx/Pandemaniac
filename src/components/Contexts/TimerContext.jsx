import React,{ useState, createContext } from 'react';

export const TimerContext = createContext()

export const TimerProvider = props => {

    let [time, setTime] = useState(0)

    return (
        <TimerContext.Provider value={[time, setTime]}>
            {props.children}
        </TimerContext.Provider>
    )
}

