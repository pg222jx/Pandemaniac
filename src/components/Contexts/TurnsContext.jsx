import React,{ useState, createContext } from 'react';

export const TurnsContext = createContext()

export const TurnsProvider = props => {

    const [amount, setAmount] = useState(0)

    return (
        <TurnsContext.Provider value={[amount, setAmount]}>
            {props.children}
        </TurnsContext.Provider>
    )
}

