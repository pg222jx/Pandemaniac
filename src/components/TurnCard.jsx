/**
 * The action card.
 *
 * @author Pernilla GÃ¶th
 * @version 1.0.0
 */


import React, { useContext } from 'react'
import { TimerContext } from './Contexts/TimerContext'
import { MovePlayerClickedContext } from './Contexts/MovePlayerClickedContext'
import { RenderStationClickedContext } from './Contexts/RenderStationClickedContext'
import { RemoveVirusContext } from './Contexts/RemoveVirusContext'
import { ActionsContext } from './Contexts/ActionsContext'
import { TurnsContext } from './Contexts/TurnsContext'
import Infection from './Infection'

const TurnCard = (props) => {

    const [time, setTime] = useContext(TimerContext)// eslint-disable-line no-unused-vars
    const [movePlayerClicked, setMovePlayerClicked] = useContext(MovePlayerClickedContext) // eslint-disable-line no-unused-vars
    const [renderStationClicked, setRenderStationClicked] = useContext(RenderStationClickedContext) // eslint-disable-line no-unused-vars
    const [removeVirusClicked, setRemoveVirusClicked] = useContext(RemoveVirusContext) // eslint-disable-line no-unused-vars
    const [amount, setAmount] = useContext(ActionsContext) // eslint-disable-line no-unused-vars
    const [turns, setTurns] = useContext(TurnsContext) // eslint-disable-line no-unused-vars

    /**
    * Converts number to minutes and seconds.
    * 
    * @returns {string} - Representing the time in minutes and seconds.
    */
    const convertTime = () => {
        const totalSeconds = time
        const minutes = Math.floor(totalSeconds/60)
        const seconds = totalSeconds - minutes * 60
        return minutes + ':' + seconds
    }

    /**
     * Sets the clicked button to true and the others to false.
     * 
     * @param {string} str The string to check.
     */
    const setClick = (str) => {
        // Setting others to false to make sure you cant do two things at same time
        if (str === 'player') {
            setMovePlayerClicked(true)
            setRenderStationClicked(false)
            setRemoveVirusClicked(false)
        } else if (str === 'station') {
            setRenderStationClicked(true)
            setMovePlayerClicked(false)
            setRemoveVirusClicked(false)
        } else if (str === 'virus') {
            setRemoveVirusClicked(true)
            setRenderStationClicked(false)
            setMovePlayerClicked(false)
        } 
    }

    return ( 
        <div className="turnCard">
            <div className="TurnsDone">
                <h4>Turns done:</h4>
                <p>{turns}</p>
                <hr/>
            </div>
            <div className="timerContainer">
                <h4>Time spent:</h4>
                <p data-testid="TimeSpent-test" className="TimeSpent">{convertTime()}</p>
                <hr/>
            </div>
            <div className="turnContainer">
                <h4 className="ActionsLeft">Actions left: {amount}</h4>
                <hr/>
                <h4>What's your next move?</h4>
                <button data-testid="value" className="Turn" id="move" onClick={() => setClick('player')}>Move one step</button>
                <button className="Turn" onClick={() => setClick('station')}>Build a science centre</button>
                <button className="Turn" id="remove" onClick={() => setClick('virus')}>Remove virus</button>
                <Infection information={props.information} spawnRandomVirus={props.spawnRandomVirus} onWin={props.onWin}/>
            </div>
        </div>
     )
}
 
export default TurnCard
