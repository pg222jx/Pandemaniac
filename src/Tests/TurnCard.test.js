import React from 'react'
import ReactDOM from 'react-dom'
import { render, screen  } from '@testing-library/react'
import { shallow, mount } from 'enzyme'
import  { MovePlayerClickedProvider } from '../components/Contexts/MovePlayerClickedContext'
import  { RenderStationClickedProvider } from '../components/Contexts/RenderStationClickedContext'
import  { TimerContext, TimerProvider } from '../components/Contexts/TimerContext'
import  { RemoveVirusProvider } from '../components/Contexts/RemoveVirusContext'
import  { TurnsProvider } from '../components/Contexts/TurnsContext'
import  { ActionsContext, ActionsProvider } from '../components/Contexts/ActionsContext'
import TurnCard from '../components/TurnCard'
import { ShownStationsProvider } from '../components/Contexts/ShownStationsContext'

function renderTurnCard(amount, setAmount, time, setTime) {
    return render(
    <RemoveVirusProvider>
        <ActionsContext.Provider value={[amount, setAmount]}>
            <TimerContext.Provider value={[time, setTime]}>
                <MovePlayerClickedProvider>
                    <RenderStationClickedProvider>
                        <TurnsProvider>
                            <ShownStationsProvider>
                                <TurnCard />
                            </ShownStationsProvider>
                        </TurnsProvider>
                    </RenderStationClickedProvider>
                </MovePlayerClickedProvider>
            </TimerContext.Provider>
        </ActionsContext.Provider>
    </RemoveVirusProvider>
    )
  }

const wrapperShallow = shallow(
    <RemoveVirusProvider>
        <ActionsProvider>
            <TimerProvider>
                <MovePlayerClickedProvider>
                    <RenderStationClickedProvider>
                        <TurnsProvider>
                        <ShownStationsProvider>
                                <TurnCard />
                            </ShownStationsProvider>
                        </TurnsProvider>
                    </RenderStationClickedProvider>
                </MovePlayerClickedProvider>
            </TimerProvider>
        </ActionsProvider>
    </RemoveVirusProvider>
)

const wrapperMount = mount(
    <RemoveVirusProvider>
        <ActionsProvider>
            <TimerProvider>
                <MovePlayerClickedProvider>
                    <RenderStationClickedProvider>
                        <TurnsProvider>
                        <ShownStationsProvider>
                                <TurnCard />
                            </ShownStationsProvider>
                        </TurnsProvider>
                    </RenderStationClickedProvider>
                </MovePlayerClickedProvider>
            </TimerProvider>
        </ActionsProvider>
    </RemoveVirusProvider>
)

describe('Testing the TurnCard component', () => {
    it('should match the snapshot', () => {
        expect(wrapperShallow.html()).toMatchSnapshot()
      })

    it('should render without crashing', () => {
        const div = document.createElement('div')
        ReactDOM.render(
            <RemoveVirusProvider>
            <ActionsProvider>
                <TimerProvider>
                    <MovePlayerClickedProvider>
                        <RenderStationClickedProvider>
                            <TurnsProvider>
                            <ShownStationsProvider>
                                <TurnCard />
                            </ShownStationsProvider>
                            </TurnsProvider>
                        </RenderStationClickedProvider>
                    </MovePlayerClickedProvider>
                </TimerProvider>
            </ActionsProvider>
        </RemoveVirusProvider>, div
        )
    })

    it('should render with the correct context output', () => {
        expect(wrapperMount.find('h4.ActionsLeft').text()).toEqual('Actions left: 2')
        expect(wrapperMount.find('p.TimeSpent').text()).toEqual('0:0')
      })

    test('so ActionContext will set a value', () => {
      renderTurnCard(0)
      expect(screen.getByText('Actions left: 0')).toBeInTheDocument()
    })

    test('so TimerContext will set a value', () => {
        renderTurnCard(null, null, 5)
        expect(screen.getByTestId('TimeSpent-test')).toHaveTextContent('5')
      })
      
    it('should have 4 buttons with className Turn', () => {
        renderTurnCard()
          expect(wrapperMount.find('button').length).toEqual(4)
          expect(wrapperMount.find('button.Turn').length).toEqual(4)
    })

    it('should have an instance of the Infection component', () => {
        expect(wrapperMount.find('Infection').length).toEqual(1)
  })
})
