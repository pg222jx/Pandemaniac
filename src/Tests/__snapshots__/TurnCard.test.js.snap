// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the TurnCard component should match the snapshot 1`] = `"<div class=\\"turnCard\\"><div class=\\"TurnsDone\\"><h4>Turns done:</h4><p>0</p><hr/></div><div class=\\"timerContainer\\"><h4>Time spent:</h4><p data-testid=\\"TimeSpent-test\\" class=\\"TimeSpent\\">0:0</p><hr/></div><div class=\\"turnContainer\\"><h4 class=\\"ActionsLeft\\">Actions left: 2</h4><hr/><h4>What&#x27;s your next move?</h4><button data-testid=\\"value\\" class=\\"Turn\\" id=\\"move\\">Move one step</button><button class=\\"Turn\\">Build a science centre</button><button class=\\"Turn\\" id=\\"remove\\">Remove virus</button><button class=\\"Turn\\" id=\\"done\\">Done with turn!</button></div></div>"`;
